src/out_ref.rs

```diff
@@ -0,0 +1,573 @@
+    /// A non-`panic!`king return from this function **guarantees that the input
+    /// slice has been (successfully) initialized**, and that it is thus then
+    /// sound to `.assume_init()`.
+    ///
+    /// It also guarantees that the returned slice does correspond to the input
+    /// slice (_e.g._, for [`crate::ReadIntoUninit`]'s safety guarantees).
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use ::core::mem::{self, MaybeUninit};
+    /// use ::uninit::prelude::*;
+    ///
+    /// let mut array = uninit_array![_; 13];
+    /// assert_eq!(
+    ///     array.as_out().copy_from_slice(b"Hello, World!"),
+    ///     b"Hello, World!",
+    /// );
+    /// // we can thus soundly `assume_init` our array:
+    /// let array = unsafe {
+    ///     mem::transmute::<
+    ///         [MaybeUninit<u8>; 13],
+    ///         [            u8 ; 13],
+    ///     >(array)
+    /// };
+    /// assert_eq!(
+    ///     array,
+    ///     *b"Hello, World!",
+    /// );
+    /// ```
+    pub
+    fn copy_from_slice (
+        mut self: Out<'out, [T]>,
+        source_slice: &'_ [T],
+    ) -> &'out mut [T]
+    where
+        T : Copy,
+    {
+        unsafe {
+            // # Safety
+            //
+            //   - Writing to `self.0` is fine since `source_slice` only
+            //     contains initialized elements.
+            //
+            //   - the `copy_nonoverlapping()` call guarantees that the buffer
+            //     has been initialized.
+            self.r()
+                .as_uninit_mut()
+                .copy_from_slice(
+                    <[MaybeUninit<T>]>::from_ref(source_slice)
+                )
+            ;
+            self.assume_init()
+        }
+    }
+
+    /// Fill the buffer with values from a function mapping indices to values.
+    ///
+    /// # Guarantees
+    ///
+    /// `factory` is guaranteed to be called in the `0 .. self.len()` order.
+    #[inline]
+    pub
+    fn fill_with (
+        mut self: Out<'out, [T]>,
+        mut factory: impl FnMut(usize) -> T,
+    ) -> &'out mut [T]
+    {
+        self.iter_out()
+            .enumerate()
+            .for_each(|(i, at_dst)| { at_dst.write(factory(i)); })
+        ;
+        unsafe {
+            // Safety: The `len` values of the buffer have been initialized
+            self.assume_init()
+        }
+    }
+
+    /// `.reborrow().into_iter()`
+    #[inline]
+    pub
+    fn iter_out<'reborrow> (self: &'reborrow mut Out<'out, [T]>)
+      -> iter::IterOut<'reborrow, T>
+    {
+        self.r().into_iter()
+    }
+
+    #[inline]
+    pub
+    fn split_at_out (self: Out<'out, [T]>, idx: usize)
+      -> (Out<'out, [T]>, Out<'out, [T]> )
+    {
+        let (left, right) = unsafe { self.as_uninit_mut() }.split_at_mut(idx);
+        (left.into(), right.into())
+    }
+}
+
+/// `Deref` into `[MaybeUninit<T>]` to get access to the slice length related
+/// getters.
+impl<'out, T : 'out> ::core::ops::Deref for Out<'out, [T]> {
```

Comment:
Maybe the type should have a `as_uninit()` method regardless. Similar to how `Vec` has `as_slice` despite deref.

Reply:
Also note that this also leads to inheritting the `as_ptr() -> *const MaybeUninit<T>` method and others, which seems unwanted (especially due to naming, the inherent `as_mut_ptr()` returns `*mut T` instead).